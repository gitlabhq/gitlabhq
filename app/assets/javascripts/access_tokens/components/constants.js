import { __, s__ } from '~/locale';

export const EVENT_ERROR = 'ajax:error';
export const EVENT_SUCCESS = 'ajax:success';
export const FORM_SELECTOR = '#js-new-access-token-form';

export const INITIAL_PAGE = 1;
export const PAGE_SIZE = 100;

const BASE_FIELDS = [
  {
    key: 'name',
    label: __('Token name'),
    isRowHeader: true,
    sortable: true,
  },
  {
    formatter(description) {
      return description ?? '-';
    },
    key: 'description',
    label: __('Description'),
    sortable: true,
  },
  {
    formatter(scopes) {
      return scopes?.length ? scopes.join(', ') : __('no scopes selected');
    },
    key: 'scopes',
    label: __('Scopes'),
    sortable: true,
  },
  {
    key: 'createdAt',
    label: s__('AccessTokens|Created'),
    sortable: true,
  },
  {
    key: 'lastUsedAt',
    label: __('Last Used'),
    sortable: true,
  },
];

const ROLE_FIELD = {
  key: 'role',
  label: __('Role'),
  sortable: true,
};

export const FIELDS = [
  ...BASE_FIELDS,
  {
    formatter(ips) {
      return ips?.length ? ips?.join(', ') : '-';
    },
    key: 'lastUsedIps',
    label: __('Last Used IPs'),
    sortable: false,
    tdClass: 'gl-break-all gl-min-w-20',
  },
  {
    key: 'expiresAt',
    label: __('Expires'),
    sortable: true,
  },
  ROLE_FIELD,
  {
    key: 'action',
    label: __('Action'),
    tdClass: '!gl-py-3',
  },
];

export const INACTIVE_TOKENS_TABLE_FIELDS = [
  ...BASE_FIELDS,
  {
    key: 'expiresAt',
    label: __('Expired'),
    sortable: true,
  },
  ROLE_FIELD,
];
