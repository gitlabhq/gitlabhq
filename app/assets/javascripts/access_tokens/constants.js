import { GlFilteredSearchToken } from '@gitlab/ui';
import { s__, __ } from '~/locale';
import {
  OPERATORS_AFTER_BEFORE,
  OPERATORS_IS,
} from '~/vue_shared/components/filtered_search_bar/constants';
import DateToken from '~/vue_shared/components/filtered_search_bar/tokens/date_token.vue';

// Token types
export const FEED_TOKEN = 'feedToken';
export const INCOMING_EMAIL_TOKEN = 'incomingEmailToken';
export const STATIC_OBJECT_TOKEN = 'staticObjectToken';

export const DEFAULT_FILTER = [
  {
    type: 'state',
    value: {
      data: 'active',
      operator: '=',
    },
  },
];

export const FILTER_OPTIONS = [
  {
    icon: 'status',
    title: s__('AccessTokens|State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AccessTokens|Active') },
      { value: 'inactive', title: s__('AccessTokens|Inactive') },
    ],
  },
  {
    icon: 'remove',
    title: s__('AccessTokens|Revoked'),
    type: 'revoked',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'true', title: __('Yes') },
      { value: 'false', title: __('No') },
    ],
  },
  {
    icon: 'history',
    title: __('Created date'),
    type: 'created',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Expiration date'),
    type: 'expires',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Last used date'),
    type: 'last_used',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
];

export const FILTER_OPTIONS_CREDENTIALS_INVENTORY = [
  {
    icon: 'key',
    title: s__('CredentialsInventory|Type'),
    type: 'filter',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      {
        value: 'personal_access_tokens',
        title: s__('CredentialsInventory|Personal access tokens'),
      },
      { value: 'ssh_keys', title: s__('CredentialsInventory|SSH keys') },
      {
        value: 'resource_access_tokens',
        title: s__('CredentialsInventory|Project and group access tokens'),
      },
      { value: 'gpg_keys', title: s__('CredentialsInventory|GPG keys') },
    ],
  },
  {
    icon: 'user',
    title: s__('CredentialsInventory|Owner type'),
    type: 'owner_type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'human', title: __('Human') },
      { value: 'service_account', title: __('Service account') },
    ],
  },
  ...FILTER_OPTIONS,
];

export const DEFAULT_SORT = { value: 'expires', isAsc: true };

export const SORT_OPTIONS = [
  {
    text: __('Created date'),
    value: 'created',
    sort: {
      asc: 'created_asc',
      desc: 'created_desc',
    },
  },
  {
    text: __('Expiration date'),
    value: 'expires',
    sort: {
      asc: 'expires_asc',
      desc: 'expires_desc',
    },
  },
  {
    text: __('Last used date'),
    value: 'last_used',
    sort: {
      asc: 'last_used_asc',
      desc: 'last_used_desc',
    },
  },
  {
    text: __('Name'),
    value: 'name',
    sort: {
      asc: 'name_asc',
      desc: 'name_desc',
    },
  },
];

export const STATISTICS_CONFIG = [
  {
    title: s__('AccessTokens|Active tokens'),
    tooltipTitle: s__('AccessTokens|Filter for active tokens'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Tokens expiring in 2 weeks'),
    tooltipTitle: s__('AccessTokens|Filter for tokens expiring in 2 weeks'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
      {
        type: 'expires',
        value: {
          data: 'DATE_HOLDER',
          operator: '<',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Revoked tokens'),
    tooltipTitle: s__('AccessTokens|Filter for revoked tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'true',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Expired tokens'),
    tooltipTitle: s__('AccessTokens|Filter for expired tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'false',
          operator: '=',
        },
      },
      {
        type: 'state',
        value: {
          data: 'inactive',
          operator: '=',
        },
      },
    ],
  },
];
