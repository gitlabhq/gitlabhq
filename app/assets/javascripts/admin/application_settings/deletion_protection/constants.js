import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  label: s__('DeletionSettings|Retention period'),
  helpText: s__(
    'DeletionSettings|Period that deleted groups and projects will remain restorable for.',
  ),
  learnMore: __('Learn more.'),
  days: __('days'),
};

export const DEL_ADJ_PERIOD_MAX_LIMIT = 90;
export const DEL_ADJ_PERIOD_MIN_LIMIT = 1;

export const DEL_ADJ_PERIOD_MAX_LIMIT_ERROR = s__(
  'DeletionSettings|Maximum deletion protection duration is 90 days.',
);

export const DEL_ADJ_PERIOD_MIN_LIMIT_ERROR = s__(
  'DeletionSettings|Minimum deletion protection duration is 1 day.',
);
