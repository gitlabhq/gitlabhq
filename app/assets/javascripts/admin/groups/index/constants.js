import groupsEmptyStateIllustration from '@gitlab/svgs/dist/illustrations/empty-state/empty-groups-md.svg?url';
import { s__, __ } from '~/locale';
import {
  SORT_LABEL_NAME,
  SORT_LABEL_CREATED,
  SORT_LABEL_UPDATED,
  SORT_LABEL_STORAGE_SIZE,
} from '~/groups_projects/constants';
import GroupsList from '~/vue_shared/components/groups_list/groups_list.vue';
import { formatGraphQLGroups } from '~/vue_shared/components/groups_list/formatter';
import ResourceListsEmptyState from '~/vue_shared/components/resource_lists/empty_state.vue';
import adminGroupsQuery from './graphql/queries/groups.query.graphql';

const baseTab = {
  formatter: (groups) =>
    formatGraphQLGroups(groups, (group) => {
      const adminPath = `/admin/groups/${group.fullPath}`;

      return {
        avatarLabelLink: adminPath,
        editPath: `${adminPath}/edit`,
      };
    }),
  listComponent: GroupsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showGroupIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  query: adminGroupsQuery,
  queryPath: 'groups',
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  variables: { active: true },
  countsQueryPath: 'active',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any active groups yet."),
    description: s__(
      'Organization|A group is a collection of several projects. If you organize your projects under a group, it works like a folder.',
    ),
    'data-testid': 'groups-empty-state',
  },
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  variables: { active: false },
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any inactive groups."),
    description: s__('Groups|Groups that are archived or pending deletion will appear here.'),
  },
};

export const SORT_OPTION_NAME = {
  value: 'name',
  text: SORT_LABEL_NAME,
};

export const SORT_OPTION_CREATED = {
  value: 'created_at',
  text: SORT_LABEL_CREATED,
};

export const SORT_OPTION_UPDATED = {
  value: 'updated_at',
  text: SORT_LABEL_UPDATED,
};

export const SORT_OPTION_STORAGE_SIZE = {
  value: 'storage_size',
  text: SORT_LABEL_STORAGE_SIZE,
};

export const SORT_OPTIONS = [
  SORT_OPTION_NAME,
  SORT_OPTION_CREATED,
  SORT_OPTION_UPDATED,
  SORT_OPTION_STORAGE_SIZE,
];

export const ADMIN_GROUPS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/groups';

export const ADMIN_GROUPS_ROUTE_NAME = 'admin-groups';

export const FIRST_TAB_ROUTE_NAMES = [ADMIN_GROUPS_ROUTE_NAME];

export const FILTERED_SEARCH_TERM_KEY = 'search';
export const FILTERED_SEARCH_NAMESPACE = 'admin-groups';
