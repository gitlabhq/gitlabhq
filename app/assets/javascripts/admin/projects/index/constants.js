import emptyStateProjectsSvgPath from '@gitlab/svgs/dist/illustrations/empty-state/empty-projects-md.svg?url';
import { __, s__ } from '~/locale';
import ProjectsList from '~/vue_shared/components/projects_list/projects_list.vue';
import ResourceListsEmptyState, {
  TYPES,
} from '~/vue_shared/components/resource_lists/empty_state.vue';
import { formatGraphQLProjects } from '~/vue_shared/components/projects_list/formatter';
import projectsQuery from './graphql/queries/projects.query.graphql';

const baseTab = {
  listComponent: ProjectsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showProjectIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  emptyStateComponentProps: {
    title: s__('Projects|No projects found.'),
    svgPath: emptyStateProjectsSvgPath,
    searchMinimumLength: 3,
    type: TYPES.filter,
  },
  queryPath: 'projects',
  formatter: (projects) =>
    formatGraphQLProjects(projects, (project) => ({
      editPath: `/admin/projects/${project.fullPath}/edit`,
      avatarLabelLink: `/admin/projects/${project.fullPath}`,
    })),
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  query: projectsQuery,
  variables: { active: true },
  queryPath: 'projects',
  countsQueryPath: 'active',
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  query: projectsQuery,
  variables: { active: false },
  queryPath: 'projects',
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    ...baseTab.emptyStateComponentProps,
    title: s__("Projects|You don't have any inactive projects."),
    description: s__('Projects|Projects that are archived or pending deletion will appear here.'),
  },
};

export const ADMIN_PROJECTS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/projects';

export const ADMIN_PROJECTS_ROUTE_NAME = 'admin-projects';
export const FIRST_TAB_ROUTE_NAMES = [ADMIN_PROJECTS_ROUTE_NAME];

export const FILTERED_SEARCH_NAMESPACE = 'admin';
export const FILTERED_SEARCH_TERM_KEY = 'search';
