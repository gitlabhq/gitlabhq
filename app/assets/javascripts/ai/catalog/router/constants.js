export const AI_CATALOG_INDEX_ROUTE = 'ai-catalog';
