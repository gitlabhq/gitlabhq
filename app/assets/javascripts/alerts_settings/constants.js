import { s__, __ } from '~/locale';

export const i18n = {
  card: {
    title: s__('AlertSettings|Active alerts'),
  },
  integrationTabs: {
    configureDetails: s__('AlertSettings|Configure details'),
    viewCredentials: s__('AlertSettings|View credentials'),
    sendTestAlert: s__('AlertSettings|Send test alert'),
  },
  integrationFormSteps: {
    selectType: {
      label: s__('AlertSettings|Select integration type'),
      enterprise: s__(
        'AlertSettings|Free versions of GitLab are limited to one integration per type. To add more, %{linkStart}upgrade your subscription%{linkEnd}.',
      ),
    },
    nameIntegration: {
      label: s__('AlertSettings|Name integration'),
      placeholder: s__('AlertSettings|Enter integration name'),
      activeToggle: __('Active'),
      error: __("Name can't be blank"),
    },
    enableIntegration: {
      label: s__('AlertSettings|Enable integration'),
      help: s__(
        'AlertSettings|A webhook URL and authorization key is generated for the integration. After you save the integration, both are visible under the “View credentials” tab.',
      ),
    },
    setupCredentials: {
      help: s__(
        'AlertSettings|Use the URL and authorization key below to configure how an external service sends alerts to GitLab. %{linkStart}How do I configure the endpoint?%{linkEnd}',
      ),
      prometheusHelp: s__(
        'AlertSettings|Use the URL and authorization key below to configure how Prometheus sends alerts to GitLab. Review the %{linkStart}GitLab documentation%{linkEnd} to learn how to configure your endpoint.',
      ),
      webhookUrl: s__('AlertSettings|Webhook URL'),
      authorizationKey: s__('AlertSettings|Authorization key'),
      reset: s__('AlertSettings|Reset Key'),
    },
    mapFields: {
      label: s__('AlertSettings|Customize alert payload mapping (optional)'),
      help: s__(
        'AlertSettings|To create a custom mapping, enter an example payload from your monitoring tool, in JSON format. Select the "Parse payload fields" button to continue.',
      ),
      prometheusHelp: s__(
        'AlertSettings|To create a custom mapping, enter a single JSON object from the array under the %{codeStart}alerts%{codeEnd} key of a payload from your Prometheus instance. Select the "Parse payload fields" button to continue.',
      ),
      placeholder: s__('AlertSettings|{ "events": [{ "application": "Name of application" }] }'),
      editPayload: s__('AlertSettings|Edit payload'),
      parsePayload: s__('AlertSettings|Parse payload fields'),
      payloadParsedSuccessMsg: s__(
        'AlertSettings|Sample payload has been parsed. You can now map the fields.',
      ),
      resetHeader: s__('AlertSettings|Reset the mapping'),
      resetBody: s__('AlertSettings|If you edit the payload, you must re-map the fields again.'),
      resetOk: s__('AlertSettings|Proceed with editing'),
      mapIntro: s__(
        'AlertSettings|You can map default GitLab alert fields to your payload keys in the dropdowns below.',
      ),
    },
    testPayload: {
      help: s__(
        'AlertSettings|Enter an example payload from your selected monitoring tool. This supports sending alerts to a GitLab endpoint.',
      ),
      placeholder: s__('AlertSettings|{ "events": [{ "application": "Name of application" }] }'),
      modalTitle: s__('AlertSettings|The form has unsaved changes'),
      modalBody: s__('AlertSettings|The form has unsaved changes. How would you like to proceed?'),
      savedAndTest: s__('AlertSettings|Save integration & send'),
      proceedWithoutSave: s__('AlertSettings|Send without saving'),
      cancel: __('Cancel'),
    },
    restKeyInfo: {
      label: s__(
        'AlertSettings|If you reset the authorization key for this project, you must update the key in every enabled alert source.',
      ),
    },
  },
  saveIntegration: s__('AlertSettings|Save integration'),
  saveAndTestIntegration: s__('AlertSettings|Save & create test alert'),
  cancelAndClose: __('Cancel and close'),
  send: __('Send'),
  copy: __('Copy'),
  integrationCreated: {
    title: s__('AlertSettings|Integration successfully saved'),
    successMsg: s__(
      'AlertSettings|GitLab has created a URL and authorization key for your integration. You can use them to set up a webhook and authorize your endpoint to send alerts to GitLab.',
    ),
    btnCaption: s__('AlertSettings|View URL and authorization key'),
  },
  changesSaved: s__('AlertsIntegrations|The integration is saved.'),
  integrationRemoved: s__('AlertsIntegrations|The integration is deleted.'),
  alertSent: s__('AlertsIntegrations|The test alert should now be visible in your alerts list.'),
  addNewIntegration: s__('AlertSettings|Add new integration'),
  settingsTabs: {
    currentIntegrations: s__('AlertSettings|Current integrations'),
    integrationSettings: s__('AlertSettings|Alert settings'),
  },
};

export const integrationSteps = {
  selectType: 'SELECT_TYPE',
  nameIntegration: 'NAME_INTEGRATION',
  enableHttpIntegration: 'ENABLE_HTTP_INTEGRATION',
  enablePrometheusIntegration: 'ENABLE_PROMETHEUS_INTEGRATION',
  customizeMapping: 'CUSTOMIZE_MAPPING',
};

export const integrationTypes = {
  none: { value: '', text: s__('AlertSettings|Select integration type') },
  http: { value: 'HTTP', text: s__('AlertSettings|HTTP Endpoint') },
  prometheus: { value: 'PROMETHEUS', text: s__('AlertSettings|Prometheus') },
};

export const typeSet = {
  http: 'HTTP',
  prometheus: 'PROMETHEUS',
};

export const integrationToDeleteDefault = { id: null, name: '' };

export const JSON_VALIDATE_DELAY = 250;

export const targetPrometheusUrlPlaceholder = 'http://prometheus.example.com/';

/**
 * Tracks snowplow event when user views alerts integration list
 */
export const trackAlertIntegrationsViewsOptions = {
  /* eslint-disable-next-line @gitlab/require-i18n-strings */
  category: 'Alert Integrations',
  action: 'view_alert_integrations_list',
};

export const mappingFields = {
  mapping: 'mapping',
  fallback: 'fallback',
};

export const tabIndices = {
  configureDetails: 0,
  viewCredentials: 1,
  sendTestAlert: 2,
};

export const testAlertModalId = 'confirmSendTestAlert';

/* Alerts integration settings constants */

export const I18N_ALERT_SETTINGS_FORM = {
  saveBtnLabel: __('Save changes'),
  introText: __('Action to take when receiving an alert. %{docsLink}'),
  introLinkText: __('Learn more.'),
  createIncident: {
    label: __('Create an incident. Incidents are created for each alert triggered.'),
  },
  incidentTemplate: {
    label: __('Incident template (optional).'),
  },
  sendEmail: {
    label: __('Send a single email notification to Owners and Maintainers for new alerts.'),
  },
  autoCloseIncidents: {
    label: __(
      'Automatically close associated incident when a recovery alert notification resolves an alert',
    ),
  },
};

export const NO_ISSUE_TEMPLATE_SELECTED = { value: '', text: __('No template selected') };
export const TAKING_INCIDENT_ACTION_DOCS_LINK =
  '/help/operations/incident_management/alerts#trigger-actions-from-alerts';
export const ISSUE_TEMPLATES_DOCS_LINK =
  '/help/user/project/description_templates#create-an-issue-template';
