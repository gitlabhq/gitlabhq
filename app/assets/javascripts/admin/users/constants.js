import { GlFilteredSearchToken } from '@gitlab/ui';

import { OPERATORS_IS } from '~/vue_shared/components/filtered_search_bar/constants';
import { s__, __, sprintf } from '~/locale';

export const I18N_USER_ACTIONS = {
  edit: __('Edit'),
  editWithName: (name) => sprintf(__('Edit user: %{user_name}'), { user_name: name }),
  userAdministration: s__('AdminUsers|User administration'),
  unlock: __('Unlock'),
  block: s__('AdminUsers|Block'),
  unblock: s__('AdminUsers|Unblock'),
  approve: s__('AdminUsers|Approve'),
  reject: s__('AdminUsers|Reject'),
  deactivate: s__('AdminUsers|Deactivate'),
  activate: s__('AdminUsers|Activate'),
  ldapBlocked: s__('AdminUsers|Cannot unblock LDAP blocked users'),
  delete: s__('AdminUsers|Delete user'),
  deleteWithContributions: s__('AdminUsers|Delete user and contributions'),
  ban: s__('AdminUsers|Ban user'),
  unban: s__('AdminUsers|Unban user'),
  trust: s__('AdminUsers|Trust user'),
  untrust: s__('AdminUsers|Untrust user'),
};

const createTokenConfig = (options) => ({
  token: GlFilteredSearchToken,
  operators: OPERATORS_IS,
  unique: true,
  ...options,
});

// This is exported so the EE version of constants.js can call it.
export const getTokenConfigs = (accessLevelOptions) => [
  createTokenConfig({
    title: s__('AdminUsers|Access level'),
    type: 'access_level',
    options: accessLevelOptions,
  }),
  createTokenConfig({
    title: __('State'),
    type: 'state',
    options: [
      { value: 'active', title: s__('AdminUsers|Active') },
      { value: 'banned', title: s__('AdminUsers|Banned') },
      { value: 'blocked', title: s__('AdminUsers|Blocked') },
      { value: 'deactivated', title: s__('AdminUsers|Deactivated') },
      { value: 'blocked_pending_approval', title: s__('AdminUsers|Pending approval') },
      { value: 'trusted', title: s__('AdminUsers|Trusted') },
      { value: 'wop', title: s__('AdminUsers|Without projects') },
    ],
  }),
  createTokenConfig({
    title: s__('AdminUsers|Two-factor authentication'),
    type: '2fa',
    options: [
      { value: 'two_factor_enabled', title: __('On') },
      { value: 'two_factor_disabled', title: __('Off') },
    ],
  }),
  createTokenConfig({
    title: __('Type'),
    type: 'type',
    options: [
      { value: 'without_bots', title: s__('AdminUsers|Humans') },
      { value: 'bots', title: s__('AdminUsers|Bots') },
      { value: 'placeholder', title: s__('UserMapping|Placeholder') },
    ],
  }),
  createTokenConfig({
    title: s__('AdminUsers|LDAP sync'),
    type: 'ldap_sync',
    options: [{ value: 'ldap_sync', title: __('True') }],
  }),
];

export const TOKEN_CONFIGS = getTokenConfigs([
  { value: 'admins', title: s__('AdminUsers|Administrator') },
  { value: 'external', title: s__('AdminUsers|External') },
]);

export const SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT = 10;

export const SOLO_OWNED_ORGANIZATIONS_EMPTY = {
  count: 0,
  nodes: [],
};
