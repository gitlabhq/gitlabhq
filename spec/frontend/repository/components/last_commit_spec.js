import Vue from 'vue';
import { GlLoadingIcon } from '@gitlab/ui';
import VueApollo from 'vue-apollo';
import { useMockInternalEventsTracking } from 'helpers/tracking_internal_events_helper';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import LastCommit from '~/repository/components/last_commit.vue';
import CommitInfo from '~/repository/components/commit_info.vue';
import SignatureBadge from '~/commit/components/signature_badge.vue';
import CiIcon from '~/vue_shared/components/ci_icon/ci_icon.vue';
import eventHub from '~/repository/event_hub';
import pathLastCommitQuery from 'shared_queries/repository/path_last_commit.query.graphql';
import pipelineCiStatusUpdatedSubscription from '~/graphql_shared/subscriptions/pipeline_ci_status_updated.subscription.graphql';
import { FORK_UPDATED_EVENT } from '~/repository/constants';
import { mockPipelineStatusUpdatedResponse, createCommitData } from '../mock_data';

Vue.use(VueApollo);

const { bindInternalEventDocument } = useMockInternalEventsTracking();

describe('Repository last commit component', () => {
  let wrapper;
  let commitData;
  let mockResolver;
  let apolloProvider;

  const findLastCommitLabel = () => wrapper.findByTestId('last-commit-id-label');
  const findHistoryButton = () => wrapper.findByTestId('last-commit-history');
  const findLoadingIcon = () => wrapper.findComponent(GlLoadingIcon);
  const findStatusBox = () => wrapper.findComponent(SignatureBadge);
  const findCommitInfo = () => wrapper.findComponent(CommitInfo);
  const findPipelineStatus = () => wrapper.findComponent(CiIcon);

  const subscriptionHandler = jest.fn().mockResolvedValue(mockPipelineStatusUpdatedResponse);

  const createComponent = (
    data = {},
    pipelineSubscriptionHandler = subscriptionHandler,
    isRealTime = true,
  ) => {
    const currentPath = 'path';

    commitData = createCommitData(data);
    mockResolver = jest.fn().mockResolvedValue(commitData);

    apolloProvider = createMockApollo([
      [pathLastCommitQuery, mockResolver],
      [pipelineCiStatusUpdatedSubscription, pipelineSubscriptionHandler],
    ]);

    wrapper = shallowMountExtended(LastCommit, {
      apolloProvider,
      propsData: { currentPath, historyUrl: '/history' },
      provide: {
        glFeatures: {
          ciPipelineStatusRealtime: isRealTime,
        },
      },
    });
  };

  it.each`
    loading  | label
    ${true}  | ${'shows'}
    ${false} | ${'hides'}
  `('$label when loading icon is $loading', async ({ loading }) => {
    createComponent();

    if (!loading) {
      await waitForPromises();
    }

    expect(findLoadingIcon().exists()).toBe(loading);
  });

  it('renders a CommitInfo component', async () => {
    createComponent();

    await waitForPromises();

    const commit = { ...commitData.project?.repository.paginatedTree.nodes[0].lastCommit };

    expect(findCommitInfo().props().commit).toMatchObject(commit);
  });

  it('renders commit widget', async () => {
    createComponent();

    await waitForPromises();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders short commit ID', async () => {
    createComponent();

    await waitForPromises();

    expect(findLastCommitLabel().text()).toBe('12345678');
  });

  describe('history button', () => {
    beforeEach(async () => {
      await createComponent();
    });

    it('renders History button with correct href', () => {
      expect(findHistoryButton().attributes('href')).toContain('/history');
    });

    it('should call trackEvent method when clicked on history button', () => {
      const { trackEventSpy } = bindInternalEventDocument(wrapper.element);
      findHistoryButton().vm.$emit('click');
      expect(trackEventSpy).toHaveBeenCalledWith(
        'click_history_control_on_blob_page',
        {},
        undefined,
      );
    });
  });

  it('hides pipeline components when pipeline does not exist', async () => {
    createComponent({ pipelineEdges: [] });

    await waitForPromises();

    expect(findPipelineStatus().exists()).toBe(false);
  });

  it('renders pipeline components when pipeline exists', async () => {
    createComponent();

    await waitForPromises();

    expect(findPipelineStatus().exists()).toBe(true);
  });

  describe('created', () => {
    it('binds `epicsListScrolled` event listener via eventHub', () => {
      jest.spyOn(eventHub, '$on').mockImplementation(() => {});

      createComponent();

      expect(eventHub.$on).toHaveBeenCalledWith(FORK_UPDATED_EVENT, expect.any(Function));
    });
  });

  describe('beforeDestroy', () => {
    it('unbinds `epicsListScrolled` event listener via eventHub', () => {
      jest.spyOn(eventHub, '$off').mockImplementation(() => {});

      createComponent();

      wrapper.destroy();

      expect(eventHub.$off).toHaveBeenCalledWith(FORK_UPDATED_EVENT, expect.any(Function));
    });
  });

  it('renders the signature HTML as returned by the backend', async () => {
    const signatureResponse = {
      __typename: 'GpgSignature',
      gpgKeyPrimaryKeyid: 'xxx',
      verificationStatus: 'VERIFIED',
    };

    createComponent({
      signature: {
        ...signatureResponse,
      },
    });

    await waitForPromises();

    expect(findStatusBox().props()).toMatchObject({ signature: signatureResponse });
  });

  describe('subscription', () => {
    it('calls subscription with correct variables', async () => {
      createComponent();

      await waitForPromises();

      expect(subscriptionHandler).toHaveBeenCalledWith({
        pipelineId: 'gid://gitlab/Ci::Pipeline/167',
      });
    });

    it('does not call the subscription when feature flag is false', async () => {
      const realTime = false;

      createComponent({}, subscriptionHandler, realTime);

      await waitForPromises();

      expect(subscriptionHandler).not.toHaveBeenCalled();
    });
  });

  describe('polling', () => {
    it('polls for last commit and pipeline data', () => {
      createComponent();

      expect(LastCommit.apollo.commit.pollInterval).toBe(30000);
    });
  });
});
