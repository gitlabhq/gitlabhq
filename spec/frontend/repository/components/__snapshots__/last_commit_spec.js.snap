// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository last commit component renders commit widget 1`] = `
<div>
  <commit-info-stub
    class="gl-hidden sm:gl-flex"
    commit="[object Object]"
  >
    <div
      class="commit-actions gl-flex gl-gap-3 gl-items-start gl-my-2"
    >
      <div
        class="gl-flex gl-h-7 gl-items-center gl-ml-5"
      >
        <ci-icon-stub
          class="gl-mr-2"
          showtooltip="true"
          status="[object Object]"
          uselink="true"
        />
      </div>
      <gl-button-group-stub
        class="gl-flex gl-items-center gl-ml-4 js-commit-sha-group"
      >
        <gl-button-stub
          buttontextclasses=""
          category="primary"
          class="dark:!gl-bg-strong gl-font-monospace"
          data-testid="last-commit-id-label"
          icon=""
          label="true"
          size="medium"
          tag="button"
          type="button"
          variant="default"
        >
          12345678
        </gl-button-stub>
        <clipboard-button-stub
          category="secondary"
          class="dark:!gl-border-l-section input-group-text"
          size="medium"
          text="123456789"
          title="Copy commit SHA"
          tooltipplacement="top"
          variant="default"
        />
      </gl-button-group-stub>
      <gl-button-stub
        buttontextclasses=""
        category="secondary"
        class="!gl-ml-0"
        data-testid="last-commit-history"
        href="/history"
        icon=""
        size="medium"
        tag="button"
        type="button"
        variant="default"
      >
        History
      </gl-button-stub>
    </div>
  </commit-info-stub>
  <collapsible-commit-info-stub
    class="!gl-border-t-0 gl-block sm:gl-hidden"
    commit="[object Object]"
    historyurl="/history"
  />
</div>
`;
